src/EventSourcing/Infrastructure/Serialization/DeSerializeEvent.php

```diff
@@ -0,0 +1,22 @@
+<?php
+
+declare(strict_types=1);
+
+namespace EventSourcingWorkshop\EventSourcing\Infrastructure\Serialization;
+
+use EventSourcingWorkshop\EventSourcing\Domain\DomainEvent;
+
+interface DeSerializeEvent
+{
+    /**
+     * @TODO do we want to enforce `class-string<RequestedEventType>`? That forces translation in upstream.
+     *
+     * @template RequestedEventType of DomainEvent
+     *
+     * @param class-string<RequestedEventType> $type
+     * @param array<non-empty-string, mixed>   $payload
+     *
+     * @return RequestedEventType
+     */
+    public function __invoke(string $type, array $payload): DomainEvent;
```

Comment:
Is the `$payload` always an array? I'm not familiar with event sourcing so that might be the case.

Reply:
Maybe `scalar|array<non-empty-string, mixed>`?

Reply:
A `string` would suffice: we read JSON from the DB most of the time 👍

Reply:
But that's my point actually: is there no way the actual input is — for instance — an integer? `42` is a valid JSON entry AFAIK.

Reply:
No, usually not possible, although I don't really care: from a `DeSerializeEvent` perspective, the input may be a heavily compressed string, for example.